### PR TITLE
Send whole sdfg with rpc

### PR DESCRIPTION
**Pull request by NoraHagmeyer**
## Description
Prepares removing the loop normalization and bringing down compile time for segformer with target cuda (with and without remote_tuning) .

- Replaces the RPCScheduler in the LoopSchedulingPass with an RPCSchedulingPass
- Sends the whole sdfg to the RPC server (still using a transformation, to be made more efficient)
- Changes the transer tuning rest endpoint to a newly added one
- Handles response of multiple region informations
- Removes the RPC Opt application

- Changes the result in the Opt Report (to be re-added later or changed to a better mechanism for tracking) regression already tracked through #908 
-  Possible conflicts with #880 

## Related Issues

- Closes #

## Type of Change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Refactor / internal change (no functional change)
- [ ] Documentation update

## Quality Checklist
*Please ensure the following are completed before requesting review:*

* [ ] Runtime-compatible (no externally visible existing method has its parameters changed, was deleted or has its exception behavior changed)
* [x] Compile-time compatible (new arguments & properties added have default values, but users need to recompile against the new version to work)

IR:
* [ ] Introduces new types into the SDFG IR
* [ ]  Changed / added properties on SDFG IR Elements (LibraryNodes, ControlFlowNodes DataFlowNodes etc.)
  *  [ ] New / modified types have serialization / deserialization support
  *  [ ] Deserialization is backward compatible (new serialization can deserialize old format and represent it in the new format if applicable)
  *  [ ] Serialized format is backward compatible (previous deserializer can read new serialized format without incorrectness (just lacking additional guarantees that are optional or did not exist before)
  *  [ ] Serializer has an option to emit the previous format

### Unit Tests
- [x] New unit tests have been added for the changes.
- [x] Existing unit tests pass locally.
- [x] Edge cases and error paths are covered.

### Integration Tests
- [x] New integration tests have been added (or existing ones updated).
- [x] Integration tests pass locally.
- [x] Interactions with other components/services are verified.

### Documentation
- [ ] Public APIs, options, and behavior changes are documented.
- [ ] README / docs site / inline comments updated where applicable.
- [ ] Changelog / release notes updated (if applicable).

### Test Coverage
- [ ] Test coverage has not decreased as a result of this change.
- [ ] Coverage report reviewed and acceptable.
- [ ] Critical / new code paths are covered.

## How Has This Been Tested?
RPC unit tests in combination with local rpc server
Transfer Tuning Integration tests in combination with local rpc server
Ran benchmarks for this code state

### CODE REVIEW

File: rpc/src/transformations/rpc_node_transform.cpp [hunk 1]
```diff
@@ -40,34 +40,45 @@ std::string RPCNodeTransform::get_node_id_str() const { return std::to_string(th
...
-        if (loopnest_idx >= 0) {
-            report_->in_outermost_loop(loopnest_idx);
+    bool applicable = false;
+    auto outermost_loops = loop_analysis.outermost_loops();
+
+    std::cerr << "Found " << outermost_loops.size() << " outermost loops in node " << get_node_id_str() << std::endl;
+
+    for (auto outermost : outermost_loops) {
+        std::cerr << "[RPC] Checking applicability for node " << get_node_id_str()
+                  << " (loopnest_index=" << loop_analysis.loop_info(outermost).loopnest_index << ")" << std::endl;
```
Comment (Moehre2) [inline rpc/src/transformations/rpc_node_transform.cpp]: Should this be `std::cerr`? This means it is also printed in Release mode. If it is only debug output `DEBUG_PRINTLN` should be used.

File: rpc/src/transformations/rpc_node_transform.cpp [hunk 2]
```diff
@@ -260,51 +319,43 @@ void RPCNodeTransform::
...
         auto recipe = opt.local_replay.value();
-        std::cout << "Applied RPC optimization sequence with speedup " << opt.metadata.speedup
-                  << " and vector distance " << opt.metadata.vector_distance << " to loopnest " << element_id << ":\n";
-
-        if (dump_steps_) {
-            print_transformation_sequence(recipe.sequence);
+        for (const auto& region_result : opt.results) {
+            std::cout << "Applied RPC optimization sequence with speedup " << region_result.metadata.speedup
+                      << " and vector distance " << region_result.metadata.vector_distance << " to loopnest "
+                      << element_id << ":\n";
```
Comment (Moehre2) [inline rpc/src/transformations/rpc_node_transform.cpp]: Again, should this go to `std::cout` - not `DEBUG_PRINTLN`? (Although this used `std::cout` before)

File: rpc/src/transformations/rpc_node_transform.cpp [hunk 3]
```diff
@@ -260,51 +319,43 @@ void RPCNodeTransform::
...
         }
-
     } else {
-        std::cout << "RPC: Applied plain SDFG with speedup " << opt.metadata.speedup << " and vector distance "
-                  << opt.metadata.vector_distance << "\n";
+        for (const auto& region_result : opt.results) {
+            std::cout << "RPC: Applied plain SDFG with speedup " << region_result.metadata.speedup
+                      << " and vector distance " << region_result.metadata.vector_distance << "\n";
```
Comment (Moehre2) [inline rpc/src/transformations/rpc_node_transform.cpp]: See above.

**Review by Moehre2 — DISMISSED** [review]
Not sure if this is intended or slipped through... Therefore, approving.

File: python/bindings/py_structured_sdfg.cpp [hunk 4]
```diff
@@ -495,6 +488,14 @@ void PyStructuredSDFG::schedule(const docc::target::TargetOptions& options) {
         dce.run(builder, analysis_manager);
     }
 
+    if (options.remote_tuning) {
+        std::shared_ptr<sdfg::passes::rpc::RpcContext> context =
+            sdfg::passes::rpc::DaisytunerRpcContext::from_docc_config();
+        sdfg::passes::scheduler::RPCSchedulingPass
+            rpc_scheduling_pass(context, options.target, options.category, nullptr);
```
Comment (ramonwirsch) [inline python/bindings/py_structured_sdfg.cpp]: Use the TargetOptions. This more future-proof than unpacking it.

File: python/bindings/py_structured_sdfg.cpp
Comment (ramonwirsch) [inline python/bindings/py_structured_sdfg.cpp:438]: code order: if we need this before rpc opt and only do LoopScheduling after, then we can move about "get target specific schedulers code" also to behind the actual rpc opt

Comment (ramonwirsch) [inline python/bindings/py_structured_sdfg.cpp:439]: btw: is this cleanup  actively harmful for cuda / rocm? it should not be. Then we can simply run that always and need less target-specific code

File: rpc/include/sdfg/passes/rpc/rpc_scheduling_pass.h [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+#include "sdfg/passes/rpc/rpc_context.h"
+
+namespace sdfg {
+namespace transformations {
+class Recorder;
+}
+namespace passes {
+namespace scheduler {
+
+class RPCSchedulingPass : public Pass {
```
Comment (ramonwirsch) [inline rpc/include/sdfg/passes/rpc/rpc_scheduling_pass.h]: Very restrictive name. Since it just hands over the entire SDFG to a configurable endpoint, this is not very "scheduling" specific.

File: rpc/src/transformations/rpc_node_transform.cpp [hunk 6]
```diff
@@ -40,34 +40,45 @@ std::string RPCNodeTransform::get_node_id_str() const { return std::to_string(th
...
+
+    for (auto outermost : outermost_loops) {
+        std::cerr << "[RPC] Checking applicability for node " << get_node_id_str()
+                  << " (loopnest_index=" << loop_analysis.loop_info(outermost).loopnest_index << ")" << std::endl;
+
+
+        auto loop_info = loop_analysis.loop_info(outermost);
+
+        if (!loop_info.has_side_effects) {
+            std::cerr << "Loop has no side effects, marking as applicable" << std::endl;
```
Comment (ramonwirsch) [inline rpc/src/transformations/rpc_node_transform.cpp]: here as well

File: rpc/src/transformations/rpc_node_transform.cpp
Comment (ramonwirsch) [inline rpc/src/transformations/rpc_node_transform.cpp:29]: What does this node represent, now that its whole-sdfg? Is it just sdfg.root()?
Then why have it at all?

File: rpc/include/sdfg/passes/rpc/rpc_scheduling_pass.h [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+private:
+    std::shared_ptr<rpc::RpcContext> rpc_context_;
+    std::string target_, category_;
```
Comment (ramonwirsch) [inline rpc/include/sdfg/passes/rpc/rpc_scheduling_pass.h]: More future proof: use TargetOptions instead and for as long as possible.

File: rpc/src/transformations/rpc_node_transform.cpp
Comment (NoraHagmeyer) [reply rpc/src/transformations/rpc_node_transform.cpp:29]: Yes, it is just root.

File: python/bindings/py_structured_sdfg.cpp
Comment (NoraHagmeyer) [reply python/bindings/py_structured_sdfg.cpp:439]: Openend an issue for evaluation. Too many moving parts, I want to make as few changes as possible to keep problems contained

File: rpc/src/transformations/rpc_node_transform.cpp
Comment (NoraHagmeyer) [reply rpc/src/transformations/rpc_node_transform.cpp:29]: We don't need a transform. I opened an issue to go from a transform to just doing it in the pass. The transform is really legacy.

**Review by ramonwirsch — DISMISSED** [review]

**Review by ramonwirsch — DISMISSED** [review]

**Review by ramonwirsch — APPROVED** [review]
